From 0f0a5098af1b57a6a26143818d014638ba1ca74e Mon Sep 17 00:00:00 2001
From: chenzhen <chenzhen@rock-chips.com>
Date: Thu, 6 Apr 2017 16:32:39 +0800
Subject: [PATCH 1/2] WORKAROUND: MALI: utgard: rk322x: add delay after
 enabling clk_gpu

This is a workaround for the stuck issue found in recovery test.

Change-Id: I203ffa19a77c52b1a717980b63ff43ac886fef8a
Signed-off-by: Zhen Chen <chenzhen@rock-chips.com>
---
 drivers/gpu/arm/mali400/mali/platform/rk30/mali_platform.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/gpu/arm/mali400/mali/platform/rk30/mali_platform.c b/drivers/gpu/arm/mali400/mali/platform/rk30/mali_platform.c
index fac3cb9..b805396 100644
--- a/drivers/gpu/arm/mali400/mali/platform/rk30/mali_platform.c
+++ b/drivers/gpu/arm/mali400/mali/platform/rk30/mali_platform.c
@@ -18,6 +18,7 @@
 #include "custom_log.h"
 
 #include <linux/workqueue.h>
+#include <linux/delay.h>
 #include <linux/kernel.h>
 #include <linux/slab.h>
 #include <linux/gfp.h>
@@ -29,6 +30,7 @@
 #include <linux/uaccess.h>
 #include <linux/cpufreq.h>
 #include <linux/of.h>
+#include <linux/rockchip/cpu.h>
 
 #include "mali_kernel_common.h"
 #include "mali_osk.h"
@@ -383,6 +385,10 @@ static _mali_osk_errcode_t mali_power_domain_control(bool bpower_off)
 			}
 
 			drv_data->power_state = true;
+
+			#define DELAY_IN_US (5000)
+			if (cpu_is_rk322x())
+				usleep_range(DELAY_IN_US, DELAY_IN_US + 100);
 		}
 	} else {
 		if (drv_data->power_state) {
-- 
2.7.4

